libfwupdplugin/fu-bios-attrs.c

```diff
@@ -0,0 +1,334 @@
+		g_debug("%s", error_local->message);
+		return FALSE;
+	}
+	fwupd_bios_attr_set_description(attr, data);
+
+	return TRUE;
+}
+
+static void
+fu_bios_attr_set_integer_attrs(FwupdBiosAttr *attr)
+{
+	g_autofree gchar *lower = NULL, *upper = NULL, *scalar = NULL;
+	g_autoptr(GError) error_local = NULL;
+	guint64 tmp;
+
+	if (!fu_bios_attr_get_key(attr, "min_value", lower, &error_local)) {
+		g_debug("%s", error_local->message);
+		return;
+	}
+	if (!fu_strtoull(lower, &tmp, 0, G_MAXUINT64, &error_local)) {
+		g_debug("%s", error_local->message);
+		return;
+	}
+	fwupd_bios_attr_set_lower_bound(attr, tmp);
+	if (!fu_bios_attr_get_key(attr, "max_value", upper, &error_local)) {
+		g_debug("%s", error_local->message);
+		return;
+	}
+	if (!fu_strtoull(upper, &tmp, 0, G_MAXUINT64, &error_local)) {
+		g_debug("%s", error_local->message);
+		return;
+	}
+	fwupd_bios_attr_set_lower_bound(attr, tmp);
+	if (!fu_bios_attr_get_key(attr, "scalar_increment", scalar, &error_local)) {
+		g_debug("%s", error_local->message);
+		return;
+	}
+	if (!fu_strtoull(scalar, &tmp, 0, G_MAXUINT64, &error_local)) {
+		g_debug("%s", error_local->message);
+		return;
+	}
+	fwupd_bios_attr_set_lower_bound(attr, tmp);
+}
+
+static void
+fu_bios_attr_set_type(FuBiosAttrs *self, FwupdBiosAttr *attr, const gchar *driver)
+{
+	g_autofree gchar *data = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	g_return_if_fail(FU_IS_BIOS_ATTRS(self));
+	g_return_if_fail(FWUPD_IS_BIOS_ATTR(attr));
+	g_return_if_fail(driver != NULL);
+
+	if (g_strcmp0(fwupd_bios_attr_get_name(attr), FWUPD_BIOS_ATTR_PENDING_REBOOT) == 0) {
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_READ_ONLY);
+		return;
+	}
+
+	/* lenovo thinklmi seems to be missing it even though it's mandatory :/ */
+	if (!fu_bios_attr_get_key(attr, "type", data, &error_local)) {
+		g_debug("%s", error_local->message);
+		if (!self->kernel_bug_shown)
+			g_warning("KERNEL BUG: %s doesn't export a 'type' attribute", driver);
+		self->kernel_bug_shown = TRUE;
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_ENUMERATION);
+		return;
+	}
+
+	if (g_strcmp0(data, "enumeration") == 0) {
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_ENUMERATION);
+		return;
+	}
+	if (g_strcmp0(data, "integer") == 0) {
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_INTEGER);
+		fu_bios_attr_set_integer_attrs(attr);
+		return;
+	}
+	if (g_strcmp0(data, "string") == 0) {
+		fwupd_bios_attr_set_kernel_type(attr, FWUPD_BIOS_ATTR_TYPE_STRING);
+		return;
+	}
+}
+
+static void
+fu_bios_attrs_populate_attribute(FuBiosAttrs *self,
+				 const gchar *driver,
+				 const gchar *path,
+				 const gchar *name)
+{
+	g_autoptr(FwupdBiosAttr) attr = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	g_return_if_fail(FU_IS_BIOS_ATTRS(self));
+
+	attr = fwupd_bios_attr_new(name, path);
+	fu_bios_attr_set_type(self, attr, driver);
+	fu_bios_attr_set_description(attr);
```

Comment:
doesn't this need checking? maybe with GError

Reply:
They're optional attributes.  I'll plumb through errors though.